QFI with the primitives

a-matsuo opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
This PR adds a newQFI class using Linear Comb. of Unitaries with the primitives.

### Details and comments
- [x] LinCombQFI

dlasecki commented on qiskit/algorithms/gradients/qfi.py:
```diff
@@ -0,0 +1,226 @@
+        self,
+        circuits: Sequence[QuantumCircuit],
+        observables: Sequence[BaseOperator | PauliSumOp],
+        parameter_values: Sequence[Sequence[float]],
```
```suggestion
        parameter_values: Sequence[Sequence[complex]],
```

dlasecki commented on qiskit/algorithms/gradients/lin_comb_qfi.py:
```diff
@@ -0,0 +1,197 @@
+        Args:
+            estimator: The estimator used to compute the gradients.
+            phase_fix: Whether to calculate the second term (phase fix) of the QFI. Default to True.
+            derivative: The type of derivative. Can be either "real", "imag", or "complex".
+                Defaults to "real".
+            run_options: Backend runtime options used for circuit execution. The order of priority is:
+                run_options in ``run`` method > gradient's default run_options > primitive's default
+                setting. Higher priority setting overrides lower priority setting.
+        """
+        super().__init__(estimator, **run_options)
+        self._phase_fix = phase_fix
+        self._derivative = derivative
+        self._gradient = LinCombEstimatorGradient(estimator, derivative="complex", **run_options)
```
It would be helpful to see more explanation how `derivative` argument is different from `derivative` keyword in `LinCombEstimatorGradient`.

dlasecki commented on qiskit/algorithms/gradients/lin_comb_qfi.py:
```diff
@@ -0,0 +1,197 @@
+        self._gradient = LinCombEstimatorGradient(estimator, derivative="complex", **run_options)
+        self._gradient_circuits = {}
+
+    def _run(
+        self,
+        circuits: Sequence[QuantumCircuit],
+        observables: Sequence[BaseOperator | PauliSumOp],
+        parameter_values: Sequence[Sequence[float | complex]],
+        parameters: Sequence[Sequence[Parameter] | None],
+        **run_options,
+    ) -> EstimatorGradientResult:
```
The method body is very long. I would suggest splitting into private functions.

dlasecki commented on test/python/algorithms/test_qfi.py:
```diff
@@ -0,0 +1,275 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019, 2021.
```
```suggestion
# (C) Copyright IBM 2019, 2022.
```

dlasecki reviewed (commented):
Some feedback.

a-matsuo commented on qiskit/algorithms/gradients/base_qfi.py:
```diff
@@ -0,0 +1,180 @@
+    def __init__(
+        self,
+        estimator: BaseEstimator,
+        run_options: dict | None = None,
```
Thank you for mentioning. Yes, I was thinking to change after this PR is merged 
https://github.com/Qiskit/qiskit-terra/pull/8755

woodsp-ibm replied:
#8755 has been approved and labelled automerge - should hopefully be merged soon.

a-matsuo commented on qiskit/algorithms/gradients/base_qfi.py:
```diff
@@ -0,0 +1,180 @@
+        # run_options in ``run`` method > gradient's default run_options > primitive's default setting.
+        run_opts = copy(self._default_run_options)
+        run_opts.update_options(**run_options)
+        job = AlgorithmJob(
+            self._run, circuits, observables, parameter_values, parameters, **run_opts.__dict__
+        )
+        job.submit()
```
https://github.com/a-matsuo/qiskit-terra/blob/07fea57d3b2264bf94e970f0b29681e02a61a24f/qiskit/algorithms/gradients/lin_comb_estimator_qfi.py#L175
We have `try ... except` in `_run` method of each subclass for the gradients and qfi.

woodsp-ibm commented on qiskit/algorithms/gradients/lin_comb_estimator_gradient.py:
```diff
@@ -44,22 +50,34 @@ class LinCombEstimatorGradient(BaseEstimatorGradient):
+    def __init__(
+        self,
+        estimator: BaseEstimator,
+        derivative_type: DerivativeType = DerivativeType.REAL,
```
The above DerivativeType enum is not included the in the Sphinx docs and hence the type of this parameter ends up not being linked in the html so you cannot navigate to the type to find out what it is/more about it. You do detail the values in the docstring below though so the information is there but maybe it would be better to have it come out in html too so its linked.

woodsp-ibm commented on qiskit/algorithms/gradients/lin_comb_estimator_gradient.py:
```diff
@@ -44,22 +50,34 @@ class LinCombEstimatorGradient(BaseEstimatorGradient):
+                ``DerivativeType.REAL``.
+                For ``DerivativeType.REAL`` we compute 2Re[(dω⟨ψ(ω)|)O(θ)|ψ(ω)〉],
+                for ``DerivativeType.IMAG`` we compute 2Im[(dω⟨ψ(ω)|)O(θ)|ψ(ω)〉], and
+                for ``DerivativeType.COMPLEX`` we compute 2(dω⟨ψ(ω)|)O(θ)|ψ(ω)〉.
```
The text here is formatted in way its easy to read as list items more or less. It all ends up, as it stands being wrapped as paragraph text so the result is not quite as nicely laid out

![image](https://user-images.githubusercontent.com/40241007/191601016-f0076130-9e96-42c5-bca8-b7dc1d961011.png)

woodsp-ibm commented on qiskit/algorithms/gradients/lin_comb_estimator_gradient.py:
```diff
@@ -44,22 +50,34 @@ class LinCombEstimatorGradient(BaseEstimatorGradient):
         circuits: Sequence[QuantumCircuit],
         observables: Sequence[BaseOperator | PauliSumOp],
-        parameter_values: Sequence[Sequence[float]],
+        parameter_values: Sequence[Sequence[float | complex]],
```
You should just be able to do Sequence[complex] here. As a typehint c[omplex permits float or int values to be used](https://peps.python.org/pep-0484/#the-numeric-tower) While float | complex is not invalid these sorts of constructs have been cleaned up in Qiskit in the past following the PEP as linked earlier.

woodsp-ibm commented on qiskit/algorithms/gradients/lin_comb_estimator_gradient.py:
```diff
@@ -124,7 +151,7 @@ def _run(
         gradients = []
         for i, result in enumerate(results):
-            gradient_ = np.zeros(len(metadata_[i]["parameters"]))
+            gradient_ = np.zeros(len(metadata_[i]["parameters"]), dtype="complex_")
```
Does this need to be complex. Aren't the results from estimator always going to be real values for the expectations?

a-matsuo replied:
Yes, it needs to be complex. When we pass observables `X-iY` (i is an imaginary coefficient), which is used in LinCombQFI the return value will be complex.

woodsp-ibm commented on qiskit/algorithms/gradients/qfi_result.py:
```diff
@@ -0,0 +1,35 @@
+class QFIResult:
+    """Result of QFI."""
+
+    qfis: list[np.ndarray]
```
Here this returns `qfis`. In the LinComb constructor it has this `estimator: The estimator used to compute the gradients.` where like many places it says `gradients`. Do we want to keep consistent with using qfi or is qfi/gradient interchangable enough?

a-matsuo replied:
The docstring in the `LinCombQFI` constructor is wrong. I copy-and-pasted from `LinCombEstimatorGradient` and I forgot to change docstrings...

woodsp-ibm commented on qiskit/algorithms/gradients/utils.py:
```diff
@@ -377,3 +379,103 @@ def _gate_gradient(gate: Gate) -> Instruction:
+        "z",
+    ]
+
+    circuit2 = transpile(circuit, basis_gates=supported_gates, optimization_level=0)
```
An Estimator receiving this circuit presumably will further transpile this circuit. I assume that is all fine - testing via the reference estimator, which uses quantum info, I don't think has such paths right where it  would do further transpilation.

woodsp-ibm commented on releasenotes/notes/add-qfi-with-primitive-86d935d19dfff1a1.yaml:
```diff
@@ -0,0 +1,13 @@
+---
+features:
+  - |
+    New Quantum Fisher Information (QFI) Algorithm using the primitives has been added. It
```
Can we add a class ref to  `LinCombQFI` so there is something in the notes the user can click on which will take them to this new class.

woodsp-ibm commented on test/python/algorithms/test_qfi.py:
```diff
@@ -0,0 +1,265 @@
+        qfi = LinCombQFI(self.estimator)
+        param_list = [[np.pi / 4], [np.pi / 2]]
+        op = SparsePauliOp.from_list([("Z", 1)])
+        with self.assertRaises(ValueError):
+            qfi.run([qc], [op], param_list)
+        with self.assertRaises(ValueError):
+            qfi.run([qc, qc], [op, op], param_list, parameters=[[a]])
+        with self.assertRaises(ValueError):
+            qfi.run([qc, qc], [op], param_list, parameters=[[a]])
+        with self.assertRaises(ValueError):
+            qfi.run([qc], [op], [[np.pi / 4, np.pi / 4]])
```
It might be nice to have these as subtests with text to describe the validation so if something fails there is more info readily in the failure messages. As subtests these would be independent so the failure of say the first one here things would still test the rest which is not the case as it stands.

a-matsuo commented:
@dlasecki @woodsp-ibm Thank you so much for reviewing the PR! I fixed the codes based on the comments. I also changed to use `options`  instead of `run_options`. I think it's ready now.

woodsp-ibm commented on qiskit/algorithms/gradients/lin_comb_estimator_gradient.py:
```diff
@@ -44,23 +50,35 @@ class LinCombEstimatorGradient(BaseEstimatorGradient):
+            derivative_type: The type of derivative. Can be either ``DerivativeType.REAL``
+                ``DerivativeType.IMAG``, or ``DerivativeType.COMPLEX``. Defaults to
+                ``DerivativeType.REAL``.
+                * ``DerivativeType.REAL`` computes :math:`2 \mathrm{Re}[(dω⟨ψ(ω)|)O(θ)|ψ(ω)〉]`.
```
This still all wraps as paragraph text 

![image](https://user-images.githubusercontent.com/40241007/192801305-14073f29-f015-4aef-9405-5dde095fd88a.png)


I recall us in the past trying to find ways to have lists come out. Not sure adding a blank line helps, at least this is my vague recollection from the past which is why we ended up with something more complex in Nature where we did a list of items https://github.com/Qiskit/qiskit-nature/blob/d0ac5d9cfeb65f5db37a9ea2972309cd84190f29/qiskit_nature/circuit/library/ansatzes/ucc.py#L142

At this point close to release I think the resultant docs are ok - it could be improved upon but I think we should get this merged as its ok and perhaps look at improving it if we want later.

a-matsuo replied:
hmm. I'm not sure why it's not working. I added a blank line and used `+` instead of `*`. If the changed one doesn't work, I'm stuck.

woodsp-ibm replied:
As I mentioned above we struggled in the past - you can see from the link what we ended up with - this is how it comes out https://qiskit.org/documentation/nature/stubs/qiskit_nature.circuit.library.UCC.html see excitations parameter (even then there are bulleted sub-lists I see which are wrapping)

a-matsuo replied:
Oh, `+` is literally displayed as `+`. I thought it's some secret command. Let me try `-`, which looks working nicely e.g. 
- qubit_converter
- num_particles
- num_spin_orbitals

woodsp-ibm replied:
You can try - that is in the main text not under parameters. It was making lists under a parameter description we struggled with - ended up using :xxxxx: as you see for items.

woodsp-ibm replied:
I did not think how it looked in the image was too bad - I know its not what you wanted but something like this can always be looked at more later with a view to improving the resultant format.

woodsp-ibm replied:
This seems the best so far and I would leave it like this. The first sentence, instead of following straight on from the parameter is split to the next line, but the list is a bullet list of items as you wanted

![image](https://user-images.githubusercontent.com/40241007/193093161-a6d1cee1-bba8-48c4-8d1b-0c409274a48b.png)

a-matsuo replied:
Cool!

woodsp-ibm commented on qiskit/algorithms/gradients/base_qfi.py:
```diff
@@ -0,0 +1,212 @@
+from .qfi_result import QFIResult
+
+
+class BaseQFI(ABC):
```
In looking at the comment above `Abstract base class of the Quantum Fisher Information (QFI) for ``Estimator``.` would we ever have a sampler version? I note we have BaseEstimatorGradient, with Estimator in the name, but estimator is not part of this class name, i.e its BaseQFI not BaseEstimatorQFI. If there was a possibility to add Sampler variant later then might this make sense?

stefan-woerner replied:
There shouldn't be a sampler/estimator QFI since QFI is a property of the parametrized circuit only.

woodsp-ibm replied:
I think its more than just the text since it has estimator as part of the constructor - Stefan said he will touch base with you as I asked him offline to elaborate on his comment.

stefan-woerner replied:
I would remove `Estimator` from the base class.

Cryoris commented on qiskit/algorithms/gradients/base_estimator_gradient.py:
```diff
@@ -103,7 +103,7 @@ def _run(
         circuits: Sequence[QuantumCircuit],
         observables: Sequence[BaseOperator | PauliSumOp],
-        parameter_values: Sequence[Sequence[float]],
+        parameter_values: Sequence[Sequence[complex]],
```
Why are these complex and not `float` anymore? 🤔

a-matsuo replied:
Actually, I'm not sure. @dlasecki suggested to change it. It's maybe used in VarQTE?
https://github.com/Qiskit/qiskit-terra/pull/8688#discussion_r970458315

dlasecki replied:
I thought that in general these might be complex? But now also not sure if the execution path would handle that.

a-matsuo replied:
I talked to @dlasecki . He said VarQTE doesn't use complex parameter values. Let's make them back to `float`

Cryoris commented on qiskit/algorithms/gradients/base_qfi.py:
```diff
@@ -0,0 +1,182 @@
+        Raises:
+            ValueError: Invalid arguments are given.
+        """
+        # Validation
```
This seems clear enough from the context, in case you'd like to remove this comment 🙂 
```suggestion
```

Cryoris commented on qiskit/algorithms/gradients/base_qfi.py:
```diff
@@ -0,0 +1,182 @@
+                )
+
+    @property
+    def options(self) -> Options:
```
Since this is just forwarding to another method, would it be simpler to just have the `options` property be abstract?

Cryoris commented on qiskit/algorithms/gradients/lin_comb_qfi.py:
```diff
@@ -0,0 +1,251 @@
+            gradient_results = [g_job.result() for g_job in gradient_jobs]
+            results = [job.result() for job in jobs]
+        except AlgorithmError as exc:
+            raise AlgorithmError("Estimator job failed.") from exc
```
```suggestion
            raise AlgorithmError("Estimator or gradient job failed.") from exc
```

Cryoris reviewed (commented):
Some minor comments, but LGTM otherwise!

Cryoris commented on qiskit/algorithms/gradients/lin_comb_qfi.py:
```diff
@@ -0,0 +1,258 @@
+        Args:
+            estimator: The estimator used to compute the QFI.
+            phase_fix: Whether to calculate the second term (phase fix) of the QFI, which is
+                Re[(dω⟨<ψ(ω)|)|ψ(ω)><ψ(ω)|(dω|ψ(ω))>]. Default to ``True``.
```
Whether it's `Re`, `Im`, or both also depends on the `derivative_type`, right?

a-matsuo replied:
Right. The documentation was strange. I changed it.

Cryoris commented on qiskit/algorithms/gradients/lin_comb_qfi.py:
```diff
@@ -0,0 +1,258 @@
+                )
+                gradient_jobs.append(gradient_job)
+            # compute the first term in the QFI
+            qfi_circuits_ = self._qfi_circuits.get(id(circuit))
```
Should this use `_circuit_key` instead of the object id?

a-matsuo replied:
Ah, I didn't know about `_circuit_key`. I will use it. Also, I will make another PR to use `_circuit_key` in other gradient as well.

Cryoris reviewed (commented):
Overall LGTM, I left some questions below 🙂

a-matsuo commented:
Thank you for the comments! I will check them.

a-matsuo commented on qiskit/algorithms/gradients/lin_comb_qfi.py:
```diff
@@ -0,0 +1,258 @@
+        # compute the phase fix
+        if self._phase_fix:
+            for gradient_result in gradient_results:
+                phase_fix_ = np.outer(
```
I'm not 100% sure, but the left most "4" in the definition of QFI is to eliminate the "4" in a denominator, right. `LinCombEstimatorGradient` with `DerivativeType.COMPLEX` calculate the real and imaginary part of the gradient, e.g. `a+bi`. `<d/d theta psi | psi>` itself is a half of `a+bi`, so it's `(a+bi)/2`. `<d/d theta psi | psi><d/d theta psi | psi>` is `(a+bi)(a-bi) / 4`. Combining the left most "4" in the definition, we get `(a+bi)(a-bi)` which we can calculate from the result of `LinCombEstimatorGradient` with `DerivativeType.COMPLEX`, and its hermitian conjugate).

Cryoris replied:
Ah yes, if we only multiply the double derivative part (the qfi_circuits) with 4 then this matches again! Thanks for the clarification 🙂

a-matsuo commented on qiskit/algorithms/gradients/lin_comb_qfi.py:
```diff
@@ -0,0 +1,258 @@
+        qfis = []
+        for i, result in enumerate(results):
+            qfi_ = np.zeros(
+                (len(metadata_[i]["parameters"]), len(metadata_[i]["parameters"])), dtype="complex_"
```
It's an alias of `np.complex123`, but it seems `np.complex128` is compatible with Python's `complex` 🤔. (I'm not very familiar with data types. I just found `np.complex_` and used it)
https://numpy.org/doc/stable/reference/arrays.scalars.html?highlight=complex_#numpy.cdouble

I think it doesn't matter which type we use in this case. Let's use `complex`.

a-matsuo commented on qiskit/algorithms/gradients/lin_comb_qfi.py:
```diff
@@ -0,0 +1,258 @@
+                )
+                gradient_jobs.append(gradient_job)
+            # compute the first term in the QFI
+            qfi_circuits_ = self._qfi_circuits.get(id(circuit))
```
Thx! `qfi_circuit_cache` makes sense. I was looking for a better name.

a-matsuo commented on qiskit/algorithms/gradients/lin_comb_qfi.py:
```diff
@@ -0,0 +1,258 @@
+                for j, param_j in enumerate(circuit.parameters):
+                    if not param_j in param_set or i > j:
+                        continue
+                    qfi_circuits.extend(grad.gradient_circuit for grad in qfi_circuits_[i, j])
```
Right, `_make_lin_comb_qfi_circuit(circuit)` will create the all qfi circuits. The code passes a subset of it which is related to parameters user specified to an estimator though.

Cryoris reviewed (approved):
Thanks for all the work @a-matsuo!
